core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/AdaptiveReorderPartitionedJoin.java

```diff
@@ -247,4 +235,98 @@ private static StreamProperties deriveStreamPropertiesRecursively(PlanNode node,
                 .collect(toImmutableList());
         return deriveStreamPropertiesWithoutActualProperties(resolvedNode, inputProperties, metadata, session);
     }
+
+    private static class BuildToProbeLocalExchangeRewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final PlanNodeId localExchangeNodeId;
+        private final Context context;
+
+        private BuildToProbeLocalExchangeRewriter(PlanNodeId localExchangeNodeId, Context context)
+        {
+            this.localExchangeNodeId = requireNonNull(localExchangeNodeId, "localExchangeNodeId is null");
+            this.context = requireNonNull(context, "context is null");
+        }
+
+        @Override
+        protected PlanNode visitPlan(PlanNode node, RewriteContext<Void> ctx)
+        {
+            return rewriteSources(this, node, context);
+        }
+
+        @Override
+        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> ctx)
+        {
+            if (node.getStep() == PARTIAL) {
+                return rewriteSources(this, node, context);
+            }
+            return visitPlan(node, ctx);
+        }
+
+        @Override
+        public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> ctx)
+        {
+            if (node.getScope().equals(LOCAL) && node.getId().equals(localExchangeNodeId)) {
+                if (node.getSources().size() == 1) {
+                    return node.getSources().getFirst();
+                }
+
+                // Add RoundRobinExchange to replace the local exchange if there are multiple sources
+                return roundRobinExchange(
+                        context.getIdAllocator().getNextId(),
+                        LOCAL,
+                        node.getSources(),
+                        node.getOutputSymbols());
+            }
+            return node;
+        }
+    }
+
+    private static class ProbeToBuildLocalExchangeRewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final Context context;
+        private final List<Symbol> buildSymbols;
+
+        private ProbeToBuildLocalExchangeRewriter(List<Symbol> buildSymbols, Context context)
+        {
+            this.buildSymbols = requireNonNull(buildSymbols, "buildSymbols is null");
+            this.context = requireNonNull(context, "context is null");
+        }
+
+        @Override
+        protected PlanNode visitPlan(PlanNode node, RewriteContext<Void> ctx)
+        {
+            return partitionedExchange(
+                    context.getIdAllocator().getNextId(),
+                    LOCAL,
+                    node,
+                    buildSymbols,
+                    Optional.empty());
+        }
+
+        @Override
+        public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> ctx)
+        {
+            // if there are multiple sources with round-robin exchange, replace it with partitioned exchange
```

Comment:
what if there is partitined exchange? Looks like we will add another one even if not needed